### ISSUE TITLE
Unable to Connect to Native app (Windows, Firefox)

### ISSUE THREAD

**Issue by squabbi** [keeweb/keeweb-connect#15]
Firefox: 88.0.1 (64bit)
KeeWeb: v1.18.6 (070573e, 2021-05-19)
KeeWeb Connect: 0.3.6

---

I've made sure I've checked the browser integration for Firefox, under the 'KeeWeb Connect' column, and I have KeeWeb open in the Browser settings while trying to connect. I've also tried reinstalling the addon.

I've check that both KeeWeb and Firefox have a firewall exemption for 'private' networks, and I've checked that my current network connection has been marked as a private network.

I'm unable to get anything from the developer tools, nothing comes up in the console, debugger or network tab. It is however debugging the .../pages/options.html page (which might be the wrong one?) I've brought up the developer tools by right-clicking on the KeeWeb Connect options page, and choosing the 'Inspect' menu item.

The console has this error:
```
Source map error: Error: NetworkError when attempting to fetch resource.
Resource URL: moz-extension://5f5c40e4-8343-4db2-a546-294e657c54c7/js/options.js
Source Map URL: jsxRuntime.module.js.map
```

I have a feeling this is preventing me from the developer console from showing the output from options.js. If you could guide me on properly debugging it, I'd be glad to :) and if you need some more information.

The error on the options page says:

> Connection state: Error
> 
> Cannot connect to KeeWeb, please check if KeeWeb is open and browser integration is enabled in settings.

It worked straight away on my Mac, also using Firefox.

**Comment by antelle** [comment]
Hi! The actual error can be probably found in devtools open on the background page of the extension. Let's see, what is there?

**Comment by firedevelop** [comment]
Hi, thanks a lot for developing this extension, but also is not working for me.
OS: Windows 10 
Browser: Chrome 
keeweb: with Yubikey and key

**Message from Chrome**
 `DevTools failed to load source map: Could not load content for chrome-extension://pikpfmjfkekaeinceagbebpfkmkdlcjk/js/jsxRuntime.module.js.map: HTTP error: status code 404, net::ERR_UNKNOWN_URL_SCHEME`

**Extension Error:**
Connection state: Error
Cannot connect to KeeWeb, please check if KeeWeb is open and browser integration is enabled in settings.


**Keeweb Dev Tool:**
`View issues
index.html:1 2021-05-22T18:17:58.803Z [storage-file] Stop watch dir C:\Users\jhon\Documents\myDatabase
index.html:1 2021-05-22T18:18:11.586Z [open:2020] File open request
index.html:1 2021-05-22T18:18:11.586Z [open:2020] Open file from storage file
index.html:1 2021-05-22T18:18:11.586Z [open:2020] Stat file
index.html:1 2021-05-22T18:18:11.586Z [storage-file] Stat C:\Users\jhon\Documents\myDatabase\myData.kdbx
index.html:1 2021-05-22T18:18:11.587Z [storage-file] Stat done C:\Users\jhon\Documents\myDatabase\myData.kdbx 1ms
index.html:1 2021-05-22T18:18:11.587Z [open:2020] Open file from storage (1621634553292, local 1621634553292)
index.html:1 2021-05-22T18:18:11.587Z [open:2020] Load from storage
index.html:1 2021-05-22T18:18:11.587Z [storage-file] Load C:\Users\jhon\Documents\myDatabase\myData.kdbx
index.html:1 2021-05-22T18:18:11.591Z [storage-file] Loaded C:\Users\jhon\Documents\myDatabase\myData.kdbx 1621634553292 3ms
index.html:1 2021-05-22T18:18:11.591Z [open:2020] Open file from content loaded from storage
index.html:1 2021-05-22T18:18:11.591Z [storage-file] Load C:\Users\jhon\Documents\myDatabase\2020.key
index.html:1 2021-05-22T18:18:11.592Z [storage-file] Loaded C:\Users\jhon\Documents\myDatabase\2020.key 1608099816000 1ms
index.html:1 Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'sha512-xxxxxxxxxxxxxxxxxxxxxxxxxxxx==' blob:". Either the 'unsafe-inline' keyword, a hash ('sha256-ICa0DhwZQJsOd/Rn0N8H6FdQ71GfNL+op2zhAQ+Y4mM='), or a nonce ('nonce-...') is required to enable inline execution.
index.html:1 Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'sha512-xxxxxxxxxxxxxxxxxxxxxxxxxxxx==' blob:". Either the 'unsafe-inline' keyword, a hash ('sha256-ZxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxTKW5/67E='), or a nonce ('nonce-...') is required to enable inline execution.
index.html:1 2021-05-22T18:18:11.594Z [chal-resp] Calculating ChalResp using a YubiKey 
Object
index.html:1 2021-05-22T18:18:11.815Z [chal-resp] YubiKey ChalResp: touch requested
index.html:1 2021-05-22T18:18:13.111Z [chal-resp] Calculated YubiKey ChalResp
index.html:1 2021-05-22T18:18:13.934Z [file] Opened file 2020: 2340ms, R=1000000, 539 kB
index.html:1 2021-05-22T18:18:13.934Z [app] Add last open file xxxxx-xxxxx-6857-df3d-xxxxxxxx 2020 file C:\Users\jhon\Documents\myDatabase\myData.kdbx 1621634553292
index.html:1 2021-05-22T18:18:14.134Z [storage-file] Watch dir C:\Users\jhon\Documents\myDatabase
﻿
​
`


Give me details if you need some info extra.
thanks

**Comment by antelle** [comment]
Hi! Yes, I need more info, see my answer above what exactly. ↑

**Comment by firedevelop** [comment]
Hi Dimitri,
after trying many times, now is working very well.
Maybe is not related, but start to work after installing ADB to manage my
Android devices debugging.

thanks for your attention.
Best regards.

**Comment by popoway** [comment]
Hi! Thanks a lot for developing the extension but It is not working for me either.

OS: macOS 11.3.1 (M1)
Browser: Chrome 91.0.4472.69 (arm64)
KeeWeb: v1.18.6 (070573e, 2021-05-19)
KeeWeb Connect: 0.3.6

DevTools Console from KeeWeb:
```
[INFO ] 2021-05-24T02:02:09.417Z [start-profiler] Electron app started in 420ms: loading app requires=3ms, single instance lock=2ms, portable check=0ms, userdata dir=0ms, defining args=12ms, setting env=0ms, app on ready=43ms, loading settings key=24ms, reading app settings=3ms, creating main window=22ms, configuring main window=1ms, restoring main window position=0ms, setting ipc handlers=7ms, setting global shortcuts=0ms, subscribing to power events=0ms, setting request handlers=0ms, main window ready=167ms, main window shown=4ms
[DEBUG] 2021-05-24T02:02:09.418Z [settings-manager] Theme changed dark
[INFO ] 2021-05-24T02:02:09.419Z [usb-listener] Starting USB listener
[DEBUG] 2021-05-24T02:02:09.419Z [native-module-connector] Starting native module host
[INFO ] 2021-05-24T02:02:09.493Z [start-profiler] App started in 270ms: fetching=108ms, pre-init=135ms, loading modules=32ms, document ready=9ms, creating app model=1ms, loading configs=17ms, initializing modules=1ms, loading remote config=0ms, checking features=0ms, initializing storage=0ms, starting usb=0ms, first view rendering=74ms, app ready event=0ms
[INFO ] 2021-05-24T02:02:10.012Z [updater] Next update check will happen in 83716s
[INFO ] 2021-05-24T02:02:10.024Z [browser-extension-connector] Started
```
DevTools Console from background page:
```
connect @ background.js:621
async function (async)
connect @ background.js:612
processMessage @ background.js:966
(anonymous) @ background.js:945
background.js:600 Connecting to KeeWeb
background.js:692 -> KW {action: "change-public-keys", extensionName: "KeeWeb Connect", version: "0.3.6", publicKey: "vqk9GfD8eeLoBEmPxWmLFtQM7Ct1XL3zxevSRC1qTXQ=", nonce: "ZGVJSsTZJzs/FH5s45I3SXlGTlp6a+4K", …}action: "change-public-keys"clientID: "II8wfTF0ZTJBoE1ZyM96t+Jhu6oJqi9n"extensionName: "KeeWeb Connect"nonce: "ZGVJSsTZJzs/FH5s45I3SXlGTlp6a+4K"publicKey: "vqk9GfD8eeLoBEmPxWmLFtQM7Ct1XL3zxevSRC1qTXQ="version: "0.3.6"__proto__: Object
background.js:1459 KeeWeb host disconnected Specified native messaging host not found.
background.js:644 KeeWeb disconnected
background.js:621 Connect error Error: Cannot connect to KeeWeb, please check if KeeWeb is open and browser integration is enabled in settings.
    at Backend.rejectPendingRequests (background.js:715)
    at TransportNativeMessaging.<anonymous> (background.js:655)
    at TransportNativeMessaging.emit (background.js:160)
    at TransportNativeMessaging.portDisconnected (background.js:1460)
    at background.js:1436
```

**Comment by antelle** [comment]
@popoway what do you see in KeeWeb logs if you run:
```
/Applications/KeeWeb.app/Contents/MacOS/util/keeweb-native-messaging-host keeweb-connect@keeweb.info
```

**Comment by antelle** [comment]
I see that it also says: `Specified native messaging host not found`. Have you enabled Chrome browser in settings? Please validate if the manifest was installed correctly by KeeWeb: https://developer.chrome.com/docs/apps/nativeMessaging/#native-messaging-host-location

**Comment by popoway** [comment]
I did not see anything in the KeeWeb logs. After validating it turned out manifest was the suspect. I have Chrome Beta, so instead of `~/Library/Application Support/Google/Chrome Beta/NativeMessagingHosts/` the manifest was installed to `~/Library/Application Support/Google/Chrome/NativeMessagingHosts/net.antelle.keeweb.keeweb_connect.json`. I created a symlink manually and it works perfectly now. Thank you very much!
